nucypher/acumen/perception.py

```diff
@@ -30,141 +33,324 @@
 from nucypher.crypto.api import keccak_digest
 
 
+class BaseFleetState:
+
+    def __str__(self):
+        if len(self) != 0:
+            # TODO: draw the icon in color, similarly to the web version?
+            return '{checksum} ⇀{nickname}↽ {icon}'.format(icon=self.nickname.icon,
+                                                           nickname=self.nickname,
+                                                           checksum=self.checksum[:7])
+        else:
+            return 'No Known Nodes'
+
+
+class ArchivedFleetState(BaseFleetState):
+
+    def __init__(self, checksum, nickname, timestamp, population):
+        self.checksum = checksum
+        self.nickname = nickname
+        self.timestamp = timestamp
+        self.population = population
+
+    def abridged_details(self):
+        return {"nickname": self.nickname.payload() if self.nickname is not None else NO_KNOWN_NODES,
+                "updated": self.timestamp.rfc2822()}
+
+
+# Assumptions we're based on:
+# - Every supplied node object, after its constructor has finished,
+#   has a ``.checksum_address`` and ``bytes()`` (metadata)
+# - checksum address or metadata does not change for the same Python object
+# - ``this_node`` (the owner of FleetSensor) may not have a checksum address initially
+#   (when the constructor is first called), but will have one at the time of the first
+#   `record_fleet_state()` call. This applies to its metadata as well.
+# - The metadata of ``this_node`` **can** change.
+# - For the purposes of the fleet state, nodes with different metadata are considered different,
+#   even if they have the same checksum address.
+
+class FleetState(BaseFleetState):
+
+    @classmethod
+    def new(cls, this_node=None, nodes=[]):
```

Comment:
mutable default

Reply:
I don't understand why it gets such a bed rep. Arguments with default values should never be mutated anyway, and arguments in general are rarely mutated (ideally), so an incorrect usage is easy to spot. And you get a more readable signature (and the type).

But in this case this parameter is actually unused, so I'll just remove it.

Reply:
Consider this example which I think demonstrates the implicit nature of unintentional mutation:

```python
def mutable_default(nodes=[]):
    nodes.append(1)
    return nodes

x = mutable_default()
print(x)
[1]

y = mutable_default()
print(y)
[1, 1]

```

Reply:
I understand how it can backfire. But `nodes.append()` when `nodes` is an argument already calls for increased attention, and when you see that `nodes` has a default (whatever that default is), it's an instant RFC. 

Adding attributes to classes outside `__init__` causes problems much more often, and they are much harder to spot, yet nobody adds `__slots__` to everything.

Reply:
> nodes.append() when nodes is an argument already calls for increased attention

Passing a mutable default suggests that the original intent is the create a new object, not reuse one.  Otherwise - why pass it in al all?

> Adding attributes to classes outside __init__ causes problems much more often, and they are much harder to spot, yet nobody adds __slots__ to everything.

This has an explicit failure mode, `AttributeError` which is easier to spot than implicit mutation.

Plus, I'm far from the only one who thinks it's a more safe, intuitive, natural convention.

Reply:
> Passing a mutable default suggests that the original intent is the create a new object, not reuse one. Otherwise - why pass it in at all?

Everything in Python is mutable, except for some simple types. `constant_sorrow` constants, for example, which we use as defaults in many places. That's really no different than using `[]`.

As for why pass it - basically 1) to simplify the signature (the `[]` default needs no explanation as opposed to `None`) 2) to simplify the type signature, 3) to avoid the `if x is None: x = []` constructs.

> This has an explicit failure mode, AttributeError which is easier to spot than implicit mutation.

Only if you have a custom `__setattr__`, or `__slots__`. 

> Plus, I'm far from the only one who thinks it's a more safe, intuitive, natural convention.

As I said, I never really understood it. Doesn't seem intuitive or natural to me. There is some degree of unsafety, but I think it does not outweigh the simpler code, especially in a codebase that's careful about mutation. Although I do wish Python had an immutable dict-like and set-like literals; `[]` can be replaced by `()` at least.